### PR TITLE
Metabase Caching

### PR DESCRIPTION
**Pull request by dogcomplex**
# Overview

This PR includes the following proposed change(s):

Minor changes to bump up caching limits for:
- https://github.com/bcgov/invasivesbc/issues/318
Metabase API connections and UI: (v1)
- https://github.com/bcgov/invasivesbc/issues/458
- https://github.com/bcgov/invasivesbc/issues/462

And multi-select error messaging, to replace the other PR (was easier to just start over and do in my current branch)
- https://github.com/bcgov/invasivesbc/issues/487

Adds Metabase querying as demoed in sprint reviews.

## Type of change

- [X] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Locally tested with multiple Metabase queries.

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have made corresponding changes to the documentation
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: api/src/paths/metabase-fetch.ts [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        throw {
+          status: 503,
+          message: 'Failed to establish metabase session'
+        };
+      }
+
+      const sessionId = metabaseSession.data.id;
+      const response = await axios({
+        method: 'post',
+        url: METABASE_URL + `/api/card/${queryId}/query/json`,
```
Comment (sdevalapurkar) [inline api/src/paths/metabase-fetch.ts]: supernit:
```suggestion
        url: `${METABASE_URL}/api/card/${queryId}/query/json`,
```

File: api/src/paths/metabase-fetch.ts [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+      // Notes for Version 2: filter data can be found here:
+      const response2 = await axios({
+        method: 'post',
+        url: METABASE_URL + `/api/card/${queryId}/query`,
+        headers: {
+          'X-Metabase-Session': sessionId,
+          'Access-Control-Allow-Origin': '*'
+        },
+        timeout: 10000
+      });
```
Comment (sdevalapurkar) [inline api/src/paths/metabase-fetch.ts]: this part of commented code, can it be removed?

File: api/src/paths/metabase-fetch.ts [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+      console.log(response2.data.json_query.query.filter);
```
Comment (sdevalapurkar) [inline api/src/paths/metabase-fetch.ts]: can just say that the filter data can be found in:
```
res.data.json_query.query.filter
```

File: app/src/components/search/MetabaseSearch.tsx [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  const databaseChangesContext = useContext(DatabaseChangesContext);
+  const [metabaseChoices, setMetabaseChoices] = useState([]);
+
+  const getMetabaseChoicesFromTrip = useCallback(async () => {
+    let docs = await databaseContext.database.find({
+      selector: {
+        _id: 'trip'
+      }
+    });
+    if (docs.docs.length > 0) {
```
Comment (sdevalapurkar) [inline app/src/components/search/MetabaseSearch.tsx]: nit: might be able to do this
```suggestion
    if (docs.docs.length) {
```

File: app/src/components/search/MetabaseSearch.tsx [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+      return { ...tripDoc, metabaseChoices: newMetabaseChoices };
+    });
+  };
+
+  const addMetabaseChoice = (newMetabase: IMetabaseChoices) => {
+    saveChoices([...metabaseChoices, newMetabase]);
+  };
+
+  const updateMetabaseChoice = (updatedMetabase: IMetabaseChoices, index: number) => {
+    let updatedMetabaseChoices = [...metabaseChoices];
```
Comment (sdevalapurkar) [inline app/src/components/search/MetabaseSearch.tsx]: can this be a const?
```suggestion
    const updatedMetabaseChoices = [...metabaseChoices];
```

File: app/src/components/search/MetabaseSearch.tsx [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    updatedMetabaseChoices[index] = updatedMetabase;
+    saveChoices([...updatedMetabaseChoices]);
+  };
+
+  const deleteMetabaseChoice = (index: number) => {
+    let copy = [...metabaseChoices];
```
Comment (sdevalapurkar) [inline app/src/components/search/MetabaseSearch.tsx]: can this be const?
```suggestion
    const copy = [...metabaseChoices];
```

File: app/src/components/search/SearchResultsList.tsx [hunk 7]
```diff
@@ -23,12 +25,18 @@ const useStyles = makeStyles((theme: Theme) => ({
...
-  const [editIds, setEditIds] = useState<any[]>([]);
+  const [edits, setEdits] = useState<any[]>([]);
   const history = useHistory();
   const classes = useStyles();
   const invasivesApi = useInvasivesApi();
+  const databaseContext = useContext(DatabaseContext);
 
-  const navigateToEditPage = (ids: any) => {
+  const navigateToEditPage = (editsList: any) => {
+    if (new Set(editsList.map((edit) => edit.subtype)).size > 1) {
```
Comment (sdevalapurkar) [inline app/src/components/search/SearchResultsList.tsx:35]: question:
what is the `new Set()` doing? is it required?

File: app/src/components/search/SearchResultsList.tsx [hunk 8]
```diff
@@ -23,12 +25,18 @@ const useStyles = makeStyles((theme: Theme) => ({
...
+      notifyError(databaseContext, `Sorry, all activites must be the same Subtype in order to Bulk Edit.`);
```
Comment (sdevalapurkar) [inline app/src/components/search/SearchResultsList.tsx]: nit:
```suggestion
      notifyError(databaseContext, 'Sorry, all activites must be the same Subtype in order to Bulk Edit.');
```

File: app/src/components/trip/TripDataControls.tsx [hunk 9]
```diff
@@ -166,6 +170,92 @@ export const TripDataControls: React.FC = (props) => {
...
+
+        if (setOfChoices.includePhotos && row.media_keys && row.media_keys.length) {
+          try {
+            const mediaResults = await invasivesApi.getMedia(row.media_keys);
+
+            mediaResults.forEach((media) => {
+              photos.push({ filepath: media.file_name, dataUrl: media.encoded_file });
+            });
+          } catch {
+            // TODO handle errors appropriately
```
Comment (sdevalapurkar) [inline app/src/components/trip/TripDataControls.tsx:205]: question:
should this be added before merge? if not, all good

File: app/src/components/trip/TripDataControls.tsx [hunk 10]
```diff
@@ -166,6 +170,92 @@ export const TripDataControls: React.FC = (props) => {
...
+              return {
+                ...existingDoc,
+                docType: DocType.REFERENCE_ACTIVITY,
+                tripID: 'trip',
+                ...row,
+                formData: row.activity_payload.form_data,
+                activityType: row.activity_type,
+                activitySubtype: row.activity_subtype,
+                geometry: row.activity_payload.geometry,
+                photos: photos
```
Comment (sdevalapurkar) [inline app/src/components/trip/TripDataControls.tsx]: nit:
```suggestion
                photos
```

File: app/src/components/trip/TripDataControls.tsx [hunk 11]
```diff
@@ -166,6 +170,92 @@ export const TripDataControls: React.FC = (props) => {
...
+              return {
+                ...existingDoc,
+                docType: DocType.REFERENCE_POINT_OF_INTEREST,
+                tripID: 'trip',
+                ...row,
+                formData: row.point_of_interest_payload.form_data,
+                pointOfInterestType: row.point_of_interest_type,
+                pointOfInterestSubtype: row.point_of_interest_subtype,
+                geometry: [...row.point_of_interest_payload.geometry],
+                photos: photos
```
Comment (sdevalapurkar) [inline app/src/components/trip/TripDataControls.tsx]: nit:
```suggestion
                photos
```

File: app/src/components/trip/TripDataControls.tsx [hunk 12]
```diff
@@ -166,6 +170,92 @@ export const TripDataControls: React.FC = (props) => {
...
+              };
+            });
+            countPois++;
+          }
+        } catch (error) {
+          // TODO handle errors appropriately
```
Comment (sdevalapurkar) [inline app/src/components/trip/TripDataControls.tsx:244]: same question here about whether it needs to be completed in this pr or not

File: app/src/hooks/useInvasivesApi.ts [hunk 13]
```diff
@@ -100,6 +104,33 @@ export const useInvasivesApi = () => {
...
   };
 
+  /**
+   * Fetch activities or points of interest returned by Metabase queries corresponding to an id.
+   *
+   * @param {metabaseQueriesSearchCriteria} metabaseQueriesSearchCriteria
+   * @return {*}  {Promise<any>}
+   */
+  const getMetabaseQueryResults = async (metabaseQueriesSearchCriteria: IMetabaseQuerySearchCriteria): Promise<any> => {
+    const { data } = await api.post(`/api/metabase-fetch`, metabaseQueriesSearchCriteria);
```
Comment (sdevalapurkar) [inline app/src/hooks/useInvasivesApi.ts]: supernit:
```suggestion
    const { data } = await api.post('/api/metabase-fetch', metabaseQueriesSearchCriteria);
```

**Review by sdevalapurkar — COMMENTED** [review]
Looks great, few minor comments/nits/questions :)

File: app/src/components/trip/TripDataControls.tsx (hunk 9 above)
Comment (dogcomplex) [reply app/src/components/trip/TripDataControls.tsx:205]: Eh, this is just replicated from previous photos/media code.  Does need solving, but so do photos in general - would say separate issue to tackle at some point when we improve photos

File: api/src/paths/metabase-fetch.ts (hunk 2 above)
Comment (dogcomplex) [reply api/src/paths/metabase-fetch.ts]: Sure, cuz now working on V2 so will won't lose track of it

File: app/src/components/search/SearchResultsList.tsx (hunk 7 above)
Comment (dogcomplex) [reply app/src/components/search/SearchResultsList.tsx:35]: Removing duplicates to just the unique subtypes.  Seemed like a simple solution - how do you like to solve that?

**Comment by dogcomplex** [comment]
Changes pushed, lemme know if you got a preferred way to count uniques in an array and otherwise ready to merge whenever

File: app/src/components/search/SearchResultsList.tsx (hunk 7 above)
Comment (sdevalapurkar) [reply app/src/components/search/SearchResultsList.tsx:35]: yeah one of the following approaches :)

Screenshot 2021-02-07 200421

**Review by sdevalapurkar — APPROVED** [review]
nicely done! 🥇

**Pull request merged**